Add KL divergence terms for Latent SDEs

lockwo opened the pull request:
Addresses https://github.com/patrick-kidger/diffrax/issues/401. Revives https://github.com/patrick-kidger/diffrax/pull/104. Based on that PR, I made the minimal requirements to get it up to current version (e.g. taking callables instead of ODE terms since we can't make these .vf becuase _broadcast_and_upcast requires that aug_y and drift(aug_y) are the same shape, but they aren't).

lockwo commented:
Before going further (there is a lot I am going to improve/polish) I wanted to check with your thoughts on the general approach of KL being terms and exposing the user to a function that converts their problem. An alternative could be something like in torchsde where it's part of the intregration method, i.e. the user flags it at integration time.

lockwo commented:
The other big question is if I could trivially expand this to pytree's rather than just arrays

patrick-kidger commented on diffrax/_kl_term.py:
```diff
@@ -0,0 +1,175 @@
+    select the right way to compute KL divergence
+    """
+    eqx.error_if(drift, not eqx.is_array(drift), "Only array drifts are supported")
+    eqx.error_if(diffusion, not eqx.is_array(diffusion), "Only array diffusion are supported")
```
Note that you need to write `x = eqx.error_if(x, ...)`; otherwise these error checks get DCE'd.

patrick-kidger reviewed (commented):
On make these terms: I don't have super strong feelings, but compared to the original PR we have now more clearly defined what a term is in Diffrax, and I think there are other points on the design space.

To be precise: given a diffeq of the form
```
dy = f(y, z) da + g(y, z) db
dz = h(y, z) da + k(y, z) db
```
then this would be represented in Diffrax as
```
terms = (
    MultiTerm(f, g),
    MultiTerm(h, k),
)
```

In general: everything inside a `MultiTerm(...)` is all applied to the same `dfoo`. For example the SDE-specific solvers consume a `MultiTerm[ODETerm, AbstractTerm]`, for the drift and diffusion.
Meanwhile the PyTree structure of terms themselves corresponds to different `dfoo` and `dbar`. For example semi-implicit Euler takes a pair of `(AbstractTerm, AbstractTerm)`, corresponding to the two components that are being evolved.

In this case there is an argument that the extra KL-divergence term should really correspond to a new `dfoo`, and that as such the correct thing to do is to instead replace `terms` with `(terms, kl_term)`, and then provide a wrapper solver which understands this alternative term structure.

patrick-kidger commented:
On the topic of Lineax: indeed, this should definitely make handling PyTrees much easier.

lockwo commented:
I think your idea makes a lot of sense, and I made a fair amount of progress on the solver wrapper approach. I think it will require a few more changes outside the new functions (had to make some changes to subsave to log the KLs at each step) and it requires 2 redundant VF calls (unless there is a way to get the VF calls back from an arbitrary solver step?).

However, the problem I encountered is fundamentally that the problem is not really a well formed diffrax ODE. What I mean by that is the drift function in Latent SDEs takes concat(y, context) and maps it to a shape of y. However, an ODETerm requires input to vf and output of vf are the same shape (or at least broadcast able), per https://github.com/patrick-kidger/diffrax/blob/main/diffrax/_term.py#L186. However, we might map (68,) -> (4,) as a simple example of a drift term. This was able to be sidestepped with the term approach (because we directly access the function rather than the `.vf`) but for the solver it doesn't work. My method to get this to work with the solver wrapper was to either change ODETerm to remove this requirement (which seems bad) or modify the args so the solver.set gets passed args=(args, context), then the user has to define their vector field function which does the concatenation of y and args[1] which works but seems oddly specific and bespoke. Wdyt?

lockwo commented:
Ok, I added a solver wrapper (the term version is still up). With the solver wrapper comes a needed way of logging the KLs, the best way I thought of was just to add it to the SaveAt (not supporting ts currently). It's a bit specific, but maybe there are plans in the future to allow arbitrary saving of solver states? Regardless, I still have to test it, but it should (mostly) work. Not just if hijacking to solver state is ideal, but it makes sense to me as a wrapper solver. Let me know if this is what you had in mind when describing a wrapper and I will formalize this into a real PR.

lockwo commented:
Ok, I polished things up. I went with a sort of hybrid approach where the users specifies the SDEs as you described, then just wraps a solver and everything works smoothly. However, I did create internal terms, in order to get an arbitrary solver to integrate through the KL computation, that was the best way I could think of to do so, but they are completely hidden from the user. I also added the example (can be modified to add more text, or remove pmap although I do like having an example with distribution especially since its painfully slow without it) and a test and updated the docs. Taking it off draft now since its a real PR.

lockwo commented:
Looking at the tests while they are running, I see a fail at 

```
diffrax/_integrate.py:776: in diffeqsolve
    if not _term_compatible(y0, args, terms, solver.term_structure):
```

which isn't aline of code I can even find in diffrax main or in my PR?

All the tests pass locally, so if they fail on GitHub I probably just have to modify some minor thing to appease the action runner

frankschae commented:
This is a very cool feature/example! It looks like one needs to specify
```python
levy_area=diffrax.BrownianIncrement
```
in `diffrax.UnsafeBrownianPath`

lockwo commented:
Thanks @frankschae , good catch!

lockwo commented:
The test failures are all just the safe map 0.4.27 stuff

patrick-kidger commented on diffrax/_integrate.py:
```diff
@@ -803,6 +804,9 @@ def _promote(yi):
                 "`UnsafeBrownianPath` cannot be used with adaptive step sizes."
             )
 
+    if isinstance(solver, KLSolver):
+        y0 = (y0, 0.0)
+        y0 = jtu.tree_map(_promote, y0)
     # Normalises time: if t0 > t1 then flip things around.
```
I don't think this is a thing we should put in `_integrate.py`. Rather, we should have some kind of
```python
def kl_foo_bar(...):
    ...
```
function, which can be used as
```python
solver, y0 = kl_foo_bar(...)
diffeqsolve(solver=solver, y0,=y0 ...)
```

lockwo replied:
Would you want y0 out of `kl_foo_bar` to be specific `KL y0` class or would a tuple be fine? Also, would we want to modify the y0 and the solver at the same time or have a KL solver and a KL y0 transformer?

patrick-kidger commented on diffrax/_solver/kl.py:
```diff
@@ -0,0 +1,299 @@
+import operator
+from typing import Optional, Tuple, Union
```
As mentioned elsewhere, use `tuple` instead of `Tuple`!

patrick-kidger commented on diffrax/_solver/kl.py:
```diff
@@ -0,0 +1,299 @@
+)
+
+
+_DiffusionTerm = Union[ControlTerm, WeaklyDiagonalControlTerm]
```
I don't think we want to specialise on specifically a weakly-diagonal kind of control.

Rather, I think any kind of control should be admissible! Then we use Lineax to do the linear solve against whatever kind of operator is returned.

You know, maybe what we should really be doing is implementing #370! That is to say, allow the `vector_field` of `diffrax.ControlTerm(vector_field, ...)` to instead return a `lx.AbstractLinearOperator`. That would then make things exceptionally easy over here.

WDYT?

(I'll hold off on doing the rest of the review yet as I can see this may change things.)

lockwo replied:
I think that makes sense, that could reduce the parsing/if elsing complexity I have by a fair amount. I will make a PR for #370 then incorporate those changes into this and we can revisit this PR at that point.

patrick-kidger reviewed (commented):
Gosh, it's taken me a long time to get around to this 😅

Thank you for your patience!

As you can see, I think there's actually a way we can make this much much simpler. :)
